File: netDxf/IO/DxfReader.Objects.cs
using System;
using System.Collections.Generic;
using System.Linq;
using netDxf.Objects;

namespace netDxf.IO
{
    internal sealed partial class DxfReader
    {
        private readonly HashSet<string> managedReactorHandles = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        private readonly List<DatabaseRecord> databaseRecords = new List<DatabaseRecord>();
        private readonly Dictionary<string, DatabaseMetadata> entityDatabaseMetadata = new Dictionary<string, DatabaseMetadata>(StringComparer.OrdinalIgnoreCase);
        private sealed class DatabaseMetadata
        {
            internal string Owner;
            internal string Extension;
            internal readonly List<string> Reactors = new List<string>();
        }
        private sealed class DatabaseRecord
        {
            internal DxfDatabaseObject Object;
            internal SourceRecordIdentity SourceIdentity;
            internal DatabaseMetadata Metadata = new DatabaseMetadata();
            internal readonly List<Tuple<string, string, bool>> Entries = new List<Tuple<string, string, bool>>();
            internal string Default;
            internal readonly List<string> ContainerReferences = new List<string>();
            internal readonly List<string> SortKeys = new List<string>();
        }
        private DatabaseRecord ReadDatabaseRecord()
        {
            SourceRecordIdentity source = this.CurrentSourceRecord;
            string codeName = this.chunk.ReadString();
            List<DxfTag> tags = new List<DxfTag>();
            this.chunk.Next();
            while (this.chunk.Code != 0)
            {
                if (this.chunk.Code != 999) tags.Add(new DxfTag(this.chunk.Code, this.chunk.Value));
                this.chunk.Next();
            }
            if (codeName == "TABLESTYLE" && this.doc.DrawingVariables.AcadVer >= netDxf.Header.DxfVersion.AutoCad2004
                && tags.FirstOrDefault(t => t.Code == 100)?.Value as string == "AcDbTableStyle")
            {
                DatabaseRecord style = this.ReadTableStyleRecord(tags);
                style.SourceIdentity = source;
                this.databaseRecords.Add(style);
                return style;
            }
            if (codeName == "TABLECONTENT")
            {
                DatabaseRecord content = this.ReadStoredTableContentRecord(tags);
                content.SourceIdentity = source;
                this.databaseRecords.Add(content);
                return content;
            }
            if (codeName == "SUN")
            {
                DatabaseRecord sun = this.ReadSunRecord(tags);
                sun.SourceIdentity = source;
                this.databaseRecords.Add(sun);
                return sun;
            }
            if (codeName == "FIELD" || codeName == "ACAD_FIELD")
            {
                DatabaseRecord field = this.ReadStoredFieldRecord(codeName, tags);
                field.SourceIdentity = source; this.databaseRecords.Add(field); return field;
            }
            if (codeName == "DIMASSOC")
            {
                DatabaseRecord association = this.ReadStoredDimAssocRecord(tags);
                association.SourceIdentity = source;
                this.databaseRecords.Add(association);
                return association;
            }
            if (codeName == "DATATABLE")
            {
                DatabaseRecord table = this.ReadDataTableRecord(tags);
                table.SourceIdentity = source;
                this.databaseRecords.Add(table);
                return table;
            }
            if (codeName == "SECTIONSETTINGS" || codeName == "SECTION_SETTINGS")
            {
                DatabaseRecord envelope = this.ReadSectionSettingsRecord(codeName, tags);
                envelope.SourceIdentity = source;
                this.databaseRecords.Add(envelope);
                return envelope;
            }
            if (codeName == "SECTIONMANAGER" || codeName == "SECTION_MANAGER")
            {
                DatabaseRecord manager = this.ReadSectionManagerRecord(codeName, tags);
                manager.SourceIdentity = source;
                this.databaseRecords.Add(manager);
                return manager;
            }
            if (codeName == "LAYER_INDEX")
            {
                DatabaseRecord envelope = this.ReadLayerIndexRecord(tags);
                envelope.SourceIdentity = source;
                this.databaseRecords.Add(envelope);
                return envelope;
            }
            if (codeName == "LAYER_FILTER" || codeName == "OBJECT_PTR")
            {
                DatabaseRecord envelope = this.ReadLayerFilterPointerRecord(codeName, tags);
                envelope.SourceIdentity = source;
                this.databaseRecords.Add(envelope);
                return envelope;
            }
            if (codeName == "XRECORD" && this.TryReadPrivateXRecord(tags, out DatabaseRecord privateRecord))
            {
                privateRecord.SourceIdentity = source;
                this.databaseRecords.Add(privateRecord);
                return privateRecord;
            }
            DatabaseRecord result = new DatabaseRecord { SourceIdentity = source };
            string handle = null;
            int payload = 0;
            for (; payload < tags.Count; payload++)
            {
                DxfTag tag = tags[payload];
                if (tag.Code == 100 || tag.Code == 1001) break;
                if (tag.Code == 5) handle = (string)tag.Value;
                else if (tag.Code == 330) result.Metadata.Owner = (string)tag.Value;
                else if (tag.Code == 102)
                {
                    string group = (string)tag.Value;
                    bool closed = false;
                    int depth = 1;
                    while (++payload < tags.Count)
                    {
                        tag = tags[payload];
                        if (tag.Code == 102)
                        {
                            string control = (string)tag.Value;
                            if (control == "}" && --depth == 0) { closed = true; break; }
                            if (control.StartsWith("{", StringComparison.Ordinal)) depth++;
                            continue;
                        }
                        if (depth == 1 && group == "{ACAD_XDICTIONARY" && tag.Code == 360) result.Metadata.Extension = (string)tag.Value;
                        if (depth == 1 && group == "{ACAD_REACTORS" && tag.Code == 330) result.Metadata.Reactors.Add((string)tag.Value);
                    }
                    if (!closed) throw new FormatException("Unterminated database control group.");
                }
            }
            if (codeName == "DICTIONARY" || codeName == "ACDBDICTIONARYWDFLT")
            {
                DxfDictionary dictionary = codeName == "DICTIONARY" ? new DxfDictionary() : new DxfDictionaryWithDefault();
                result.Object = dictionary;
                dictionary.IsHardOwner = false;
                string pendingName = null;
                for (int i = payload; i < tags.Count; i++)
                {
                    DxfTag tag = tags[i];
                    if (tag.Code == 1001) { this.ReadDatabaseXData(dictionary, tags, i); break; }
                    switch (tag.Code)
                    {
                        case 280: dictionary.IsHardOwner = (short)tag.Value != 0; break;
                        case 281: dictionary.Cloning = (DictionaryCloningFlags)(short)tag.Value; break;
                        case 3:
                            if (pendingName != null) throw new FormatException("Dictionary name has no associated object handle.");
                            pendingName = this.DecodeEncodedNonAsciiCharacters((string)tag.Value); break;
                        case 350: case 360:
                            if (pendingName == null) throw new FormatException("Dictionary handle has no associated name.");
                            result.Entries.Add(Tuple.Create(pendingName, (string)tag.Value, tag.Code == 360)); pendingName = null; break;
                        case 340: result.Default = (string)tag.Value; break;
                    }
                }
                if (pendingName != null) throw new FormatException("Dictionary name has no associated object handle.");
            }
            else if (codeName == "XRECORD")
            {
                DxfXRecord record = new DxfXRecord(); result.Object = record;
                if (payload < tags.Count && tags[payload].Code == 100 && (string)tags[payload].Value == "AcDbXrecord") payload++;
                else throw new FormatException("XRECORD requires AcDbXrecord subclass data.");
                if (payload < tags.Count && tags[payload].Code == 280) record.Cloning = (DictionaryCloningFlags)(short)tags[payload++].Value;
                for (; payload < tags.Count; payload++)
                {
                    DxfTag tag = tags[payload];
                    if (tag.Code == 1001) { this.ReadDatabaseXData(record, tags, payload); break; }
                    if (tag.ValueType == DxfTagValueType.String) tag = new DxfTag(tag.Code, this.DecodeEncodedNonAsciiCharacters((string)tag.Value));
                    record.AddLoadedData(tag);
                }
            }
            else if (codeName == "DICTIONARYVAR")
            {
                DxfDictionaryVariable variable = new DxfDictionaryVariable(); result.Object = variable;
                for (int i = payload; i < tags.Count; i++)
                {
                    if (tags[i].Code == 280) variable.Schema = (short)tags[i].Value;
                    else if (tags[i].Code == 1) variable.Value = this.DecodeEncodedNonAsciiCharacters((string)tags[i].Value);
                    else if (tags[i].Code == 1001) { this.ReadDatabaseXData(variable, tags, i); break; }
                }
            }
            else if (codeName == "ACDBPLACEHOLDER")
            {
                result.Object = new DxfPlaceholder();
                for (int i = payload; i < tags.Count; i++)
                    if (tags[i].Code == 1001) { this.ReadDatabaseXData(result.Object, tags, i); break; }
            }
            else if (!this.ReadStoredEnvelopePayload(result, codeName, tags, payload) && !this.ReadContainerPayload(result, codeName, tags, payload) && !this.ReadGeoDataPayload(result, codeName, tags, payload) && !this.ReadOutputSettingsPayload(result, codeName, tags, payload) && !this.ReadMLeaderStylePayload(result, codeName, tags, payload) && !this.ReadLightListPayload(result, codeName, tags, payload)) result.Object = new DxfOpaqueObject(codeName, tags.Skip(payload).ToList());
            result.Object.Handle = handle;
            this.databaseRecords.Add(result);
            return result;
        }
        private void ReadDatabaseXData(DxfObject target, List<DxfTag> tags, int start)
        {
            XData data = null;
            for (int i = start; i < tags.Count; i++)
            {
                DxfTag tag = tags[i];
                if (tag.Code == 1001)
                {
                    data = new XData(this.GetApplicationRegistry(this.DecodeEncodedNonAsciiCharacters((string)tag.Value)));
                    target.XData.Add(data);
                }
                else
                {
                    if (data == null || tag.Code < 1000 || tag.Code > 1071) throw new FormatException("Invalid object XData.");
                    object value = tag.Value;
                    if (value is string text && tag.Code != 1005) value = this.DecodeEncodedNonAsciiCharacters(text);
                    data.XDataRecord.Add(new XDataRecord((XDataCode)tag.Code, value));
                }
            }
        }
        private DictionaryObject ReadDictionaryDatabaseRecord()
        {
            DatabaseRecord record = this.ReadDatabaseRecord();
            DxfDictionary typed = (DxfDictionary)record.Object;
            DictionaryObject legacy = new DictionaryObject(null) { Handle = typed.Handle, IsHardOwner = typed.IsHardOwner, Cloning = typed.Cloning };
            foreach (Tuple<string, string, bool> entry in record.Entries)
                if (!legacy.Entries.ContainsKey(entry.Item2)) legacy.Entries.Add(entry.Item2, entry.Item1);
            legacy.XData.AddRange(typed.XData.Values);
            return legacy;
        }
        private XRecord ReadXRecordDatabaseRecord()
        {
            DatabaseRecord record = this.ReadDatabaseRecord();
            if (!(record.Object is DxfXRecord typed)) return null; // Private records have no legacy layer-state projection.
            XRecord legacy = new XRecord { Handle = typed.Handle, OwnerHandle = record.Metadata.Owner, Flags = typed.Cloning };
            foreach (DxfTag tag in typed.Data) legacy.Entries.Add(new XRecordEntry(tag.Code, tag.Value));
            return legacy;
        }
        private void ImportDatabaseObjects()
        {
            foreach (DatabaseRecord record in this.databaseRecords) this.RecordSourceObject(record.Object, record.SourceIdentity);
            this.ValidateSourceIdentityDeclarations();
            if (this.databaseRecords.Count == 0) { this.ResolveSunReferences(); this.ResolveOutputSettingsReferences(); return; }
            // Reserve source identities before lazily creating the document's temporary root.
            foreach (DatabaseRecord record in this.databaseRecords)
                if (long.TryParse(record.Object.Handle, System.Globalization.NumberStyles.AllowHexSpecifier, System.Globalization.CultureInfo.InvariantCulture, out long sourceHandle) && sourceHandle >= this.doc.NumHandles && sourceHandle < long.MaxValue)
                    this.doc.NumHandles = sourceHandle + 1;
            DxfObjectDatabase database = this.doc.Objects;
            foreach (DatabaseRecord record in this.databaseRecords)
            {
                if (record.Object is DxfXRecord xrecord) foreach (DxfTag tag in xrecord.Data) database.ReserveUnresolvedReference(tag);
                if (record.Object is DxfOpaqueObject opaque) foreach (DxfTag tag in opaque.Tags) database.ReserveUnresolvedReference(tag);
                if (record.Object is DxfStoredTableContent content) foreach (DxfTag tag in content.Payload) database.ReserveUnresolvedReference(tag);
                if (record.Object is DxfStoredSectionManager manager) foreach (DxfTag tag in manager.Tags) database.ReserveUnresolvedReference(tag);
                if (record.Object is DxfTableStyle style) foreach (DxfTag tag in style.Tags) database.ReserveUnresolvedReference(tag);
                if (record.Object is DxfStoredField field) foreach (DxfTag tag in field.Payload) database.ReserveUnresolvedReference(tag);
                foreach (XData data in record.Object.XData.Values)
                    foreach (XDataRecord tag in data.XDataRecord)
                        if (tag.Code == XDataCode.DatabaseHandle) database.ReserveUnresolvedReference(new DxfTag(1005, tag.Value));
            }
            DatabaseRecord root = this.databaseRecords.FirstOrDefault(r => r.Object.Handle == this.namedDictionary?.Handle);
            HashSet<string> managed = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            if (this.layerStateManagerDictionaryHandle != null && this.dictionaries.TryGetValue(this.layerStateManagerDictionaryHandle, out DictionaryObject layerManager))
            {
                managed.Add(layerManager.Handle);
                foreach (string child in layerManager.Entries.Keys)
                {
                    managed.Add(child);
                    if (this.dictionaries.TryGetValue(child, out DictionaryObject states))
                        foreach (string stateHandle in states.Entries.Keys) managed.Add(stateHandle);
                }
            }
            if (root != null) database.ReplaceRoot((DxfDictionary)root.Object);
            foreach (DatabaseRecord record in this.databaseRecords)
            {
                if (record == root || managed.Contains(record.Object.Handle)) continue;
                DxfObject existing = this.doc.GetObjectByHandle(record.Object.Handle);
                if (existing != null)
                {
                    bool collection = record.Object is DxfDictionary && root != null
                        && record.Metadata.Owner == root.Object.Handle
                        && this.namedDictionary.Entries.TryGetValue(record.Object.Handle, out string collectionName)
                        && (collectionName == DxfObjectCode.GroupDictionary && existing == this.doc.Groups
                            || collectionName == DxfObjectCode.LayoutDictionary && existing == this.doc.Layouts
                            || collectionName == DxfObjectCode.MLineStyleDictionary && existing == this.doc.MlineStyles
                            || collectionName == DxfObjectCode.ImageDefDictionary && existing == this.doc.ImageDefinitions
                            || collectionName == DxfObjectCode.UnderlayDgnDefinitionDictionary && existing == this.doc.UnderlayDgnDefinitions
                            || collectionName == DxfObjectCode.UnderlayDwfDefinitionDictionary && existing == this.doc.UnderlayDwfDefinitions
                            || collectionName == DxfObjectCode.UnderlayPdfDefinitionDictionary && existing == this.doc.UnderlayPdfDefinitions);
                    if (!collection && !(record.Object is DxfXRecord && existing is LayerState)) throw new FormatException("Duplicate database identity: " + record.Object.Handle);
                    // These collections were constructed from the named dictionary's
                    // source handles and intentionally represent its DICTIONARY records.
                    // LayerState conversions do not preserve source identity.
                    if (collection) this.RecordSourceObject(existing, record.SourceIdentity);
                    managed.Add(record.Object.Handle); continue;
                }
                database.Register(record.Object, true);
            }
            foreach (DatabaseRecord record in this.databaseRecords)
            {
                if (managed.Contains(record.Object.Handle)) continue;
                DxfDatabaseObject item = record.Object;
                if (record != root && record.Metadata.Owner != null && record.Metadata.Owner != "0")
                {
                    item.Owner = this.GetObjectBySourceHandle(record.Metadata.Owner);
                    if (item.Owner == null) throw new FormatException("Unresolved database owner: " + record.Metadata.Owner);
                }
            }
            foreach (DatabaseRecord record in this.databaseRecords)
            {
                if (managed.Contains(record.Object.Handle)) continue;
                DxfDatabaseObject item = record.Object;
                this.ResolveContainerReferences(record);
                if (item is DxfDictionary dictionary)
                    foreach (Tuple<string, string, bool> entry in record.Entries)
                    {
                        if (record == root && DxfObjectDatabase.IsReservedName(entry.Item1)) continue;
                        DxfObject target = this.GetObjectBySourceHandle(entry.Item2);
                        if (target == null) throw new FormatException("Unresolved dictionary entry: " + entry.Item1 + " -> " + entry.Item2);
                        dictionary.AddLoaded(entry.Item1, target, entry.Item3);
                    }
                if (item is DxfDictionaryWithDefault fallback && record.Default != null && record.Default != "0")
                {
                    fallback.Default = this.GetObjectBySourceHandle(record.Default);
                    if (fallback.Default == null) throw new FormatException("Unresolved dictionary default: " + record.Default);
                }
                this.ApplyDatabaseMetadata(item, record.Metadata);
            }
            foreach (KeyValuePair<string, DatabaseMetadata> pair in this.entityDatabaseMetadata)
            {
                DxfObject target = this.GetObjectBySourceHandle(pair.Key);
                if (target != null) this.ApplyDatabaseMetadata(target, pair.Value);
            }
            this.ResolveStoredDimAssocReferences();
            this.ResolveTableStyleReferences();
            this.ResolveStoredTableContentReferences();
            this.ResolveSunReferences();
            this.ResolveStoredFields();
            this.ResolveDataTableReferences();
            this.ResolveLayerIndexReferences();
            this.ResolveSectionSettingsReferences();
            this.ResolveSectionManagerReferences();
            this.ResolveDeclaredOwnership();
            this.ResolveGeoDataHosts();
            this.ResolveOutputSettingsReferences();
            this.ResolveLightListReferences();
        }
        private void ApplyDatabaseMetadata(DxfObject item, DatabaseMetadata metadata)
        {
            if (!string.IsNullOrEmpty(metadata.Extension) && metadata.Extension != "0")
            {
                item.ExtensionDictionary = this.GetObjectBySourceHandle(metadata.Extension) as DxfDictionary;
                if (item.ExtensionDictionary == null)
                {
                    // Existing layer-state collections have their own extension-dictionary writer.
                    if (!ReferenceEquals(item, this.doc.Layers) || metadata.Extension != this.layerStateManagerDictionaryHandle || !this.IsAcceptedSourceDictionary(metadata.Extension))
                        throw new FormatException("Unresolved extension dictionary: " + metadata.Extension);
                }
                else if (item.ExtensionDictionary.Owner != item) throw new FormatException("Extension dictionary owner mismatch: " + metadata.Extension);
            }
            foreach (string handle in metadata.Reactors)
            {
                DxfObject target = this.GetObjectBySourceHandle(handle);
                if (target != null && !item.PersistentReactors.Contains(target)) item.PersistentReactors.Add(target);
                else if (target == null && handle != "0" && !this.managedReactorHandles.Contains(handle))
                    throw new FormatException("Unresolved persistent reactor: " + handle);
            }
        }
    }
}

File: javascript/netDxf/IO/DxfReader.Objects.js
// Copyright (c) netDxf contributors. MIT License; see package LICENSE.
// Original OBJECTS dispatch/import. Callers provide the same physical stream and legacy table state.
import * as api from '../../index.js';
import * as payload from '../../runtime/DatabasePayloadIO.js';
import { DatabaseRecord } from '../../runtime/DatabaseIOContext.js';
import { DictionaryObject } from '../Objects/DictionaryObject.js';
import { XRecord } from '../Objects/XRecord.js';
import { XRecordEntry } from '../Objects/XRecordEntry.js';
import { DecodeDxfText } from '../../runtime/DxfStringEncoding.js';
import { OrdinalIgnoreCaseKey } from '../../runtime/Collections.js';
import { SourceHandle } from './DxfReader.SourceIdentity.js';
import { ReadMLeaderStylePayload } from '../../runtime/MLeaderStyleIO.js';
import { FormatException,InvalidCastException } from '../../runtime/Errors.js';
const key=value=>value==null?null:OrdinalIgnoreCaseKey(value);
export function ReadDatabaseRecord(context){
  const source=context.CurrentSourceRecord,chunk=context.Chunk,codeName=chunk.ReadString(),tags=[];
  chunk.Next();while(chunk.Code!==0){if(chunk.Code!==999)tags.push(new api.DxfTag(chunk.Code,chunk.Value));chunk.Next();}
  const complete=record=>{record.SourceIdentity=source;context.databaseRecords.Add(record);return record;};
  if(codeName==='TABLESTYLE'&&context.Document.DrawingVariables.AcadVer>=14&&tags.find(t=>t.Code===100)?.Value==='AcDbTableStyle')return complete(payload.ReadTableStyleRecord(context,tags));
  const special={CELLSTYLEMAP:'ReadStoredCellStyleMapRecord',TABLEGEOMETRY:'ReadStoredTableGeometryRecord',TABLECONTENT:'ReadStoredTableContentRecord',SUNSTUDY:'ReadStoredSunStudyRecord',SUN:'ReadSunRecord',DIMASSOC:'ReadStoredDimAssocRecord',DATATABLE:'ReadDataTableRecord',LAYER_INDEX:'ReadLayerIndexRecord'};
  if(Object.hasOwn(special,codeName))return complete(payload[special[codeName]](context,tags));
  if(codeName==='FIELD'||codeName==='ACAD_FIELD')return complete(payload.ReadStoredFieldRecord(context,codeName,tags));
  if(codeName==='SECTIONSETTINGS'||codeName==='SECTION_SETTINGS')return complete(payload.ReadSectionSettingsRecord(context,codeName,tags));
  if(codeName==='SECTIONMANAGER'||codeName==='SECTION_MANAGER')return complete(payload.ReadSectionManagerRecord(context,codeName,tags));
  if(codeName==='LAYER_FILTER'||codeName==='OBJECT_PTR')return complete(payload.ReadLayerFilterPointerRecord(context,codeName,tags));
  if(codeName==='XRECORD'){const output={};if(payload.TryReadPrivateXRecord(context,tags,output))return complete(output.value);}
  const result=new DatabaseRecord();result.SourceIdentity=source;let handle=null,start=0;
  for(;start<tags.length;start++){
    let tag=tags[start];if(tag.Code===100||tag.Code===1001)break;
    if(tag.Code===5)handle=tag.Value;
    else if(tag.Code===330)result.Metadata.Owner=tag.Value;
    else if(tag.Code===102){
      const group=tag.Value;let closed=false,depth=1;
      while(++start<tags.length){tag=tags[start];
        if(tag.Code===102){const control=tag.Value;if(control==='}'&&--depth===0){closed=true;break;}if(control.startsWith('{'))depth++;continue;}
        if(depth===1&&group==='{ACAD_XDICTIONARY'&&tag.Code===360)result.Metadata.Extension=tag.Value;
        if(depth===1&&group==='{ACAD_REACTORS'&&tag.Code===330)result.Metadata.Reactors.push(tag.Value);
      }
      if(!closed)throw new FormatException('Unterminated database control group.');
    }
  }
  if(codeName==='DICTIONARY'||codeName==='ACDBDICTIONARYWDFLT'){
    const dictionary=codeName==='DICTIONARY'?new api.DxfDictionary():new api.DxfDictionaryWithDefault();result.Object=dictionary;dictionary.IsHardOwner=false;let name=null;
    for(let i=start;i<tags.length;i++){const tag=tags[i];if(tag.Code===1001){context.ReadDatabaseXData(dictionary,tags,i);break;}
      switch(tag.Code){
        case 280:dictionary.IsHardOwner=tag.Value!==0;break;case 281:dictionary.Cloning=tag.Value;break;
        case 3:if(name!==null)throw new FormatException('Dictionary name has no associated object handle.');name=DecodeDxfText(tag.Value);break;
        case 350:case 360:if(name===null)throw new FormatException('Dictionary handle has no associated name.');result.Entries.push([name,tag.Value,tag.Code===360]);name=null;break;
        case 340:result.Default=tag.Value;break;
      }
    }
    if(name!==null)throw new FormatException('Dictionary name has no associated object handle.');
  }else if(codeName==='XRECORD'){
    const record=new api.DxfXRecord();result.Object=record;
    if(start<tags.length&&tags[start].Code===100&&tags[start].Value==='AcDbXrecord')start++;else throw new FormatException('XRECORD requires AcDbXrecord subclass data.');
    if(start<tags.length&&tags[start].Code===280)record.Cloning=tags[start++].Value;
    for(;start<tags.length;start++){let tag=tags[start];if(tag.Code===1001){context.ReadDatabaseXData(record,tags,start);break;}if(tag.ValueType===api.DxfTagValueType.String)tag=new api.DxfTag(tag.Code,DecodeDxfText(tag.Value));record.AddLoadedData(tag);}
  }else if(codeName==='DICTIONARYVAR'){
    const variable=new api.DxfDictionaryVariable();result.Object=variable;
    for(let i=start;i<tags.length;i++){const tag=tags[i];if(tag.Code===280)variable.Schema=tag.Value;else if(tag.Code===1)variable.Value=DecodeDxfText(tag.Value);else if(tag.Code===1001){context.ReadDatabaseXData(variable,tags,i);break;}}
  }else if(codeName==='ACDBPLACEHOLDER'){
    result.Object=new api.DxfPlaceholder();for(let i=start;i<tags.length;i++)if(tags[i].Code===1001){context.ReadDatabaseXData(result.Object,tags,i);break;}
  }else{
    const handlers=[payload.ReadStoredEnvelopePayload,payload.ReadContainerPayload,payload.ReadGeoDataPayload,payload.ReadOutputSettingsPayload,ReadMLeaderStylePayload,payload.ReadLightListPayload];
    if(!handlers.some(read=>read(context,result,codeName,tags,start)))result.Object=new api.DxfOpaqueObject(codeName,tags.slice(start));
  }
  result.Object.Handle=handle;context.databaseRecords.Add(result);return result;
}
export function ReadDictionaryDatabaseRecord(context){
  const record=ReadDatabaseRecord(context),typed=record.Object;
  if(!(typed instanceof api.DxfDictionary))throw new InvalidCastException();
  const legacy=new DictionaryObject(null);legacy.Handle=typed.Handle;legacy.IsHardOwner=typed.IsHardOwner;legacy.Cloning=typed.Cloning;
  for(const [name,handle]of record.Entries)if(!legacy.Entries.ContainsKey(handle))legacy.Entries.Add(handle,name);
  legacy.XData.AddRange(typed.XData.Values);return legacy;
}
export function ReadXRecordDatabaseRecord(context){
  const record=ReadDatabaseRecord(context),typed=record.Object;if(!(typed instanceof api.DxfXRecord))return null;
  const legacy=new XRecord();legacy.Handle=typed.Handle;legacy.OwnerHandle=record.Metadata.Owner;legacy.Flags=typed.Cloning;
  for(const tag of typed.Data)legacy.Entries.Add(new XRecordEntry(tag.Code,tag.Value));return legacy;
}
export function ApplyDatabaseMetadata(context,item,metadata){
  if(metadata.Extension!==null&&metadata.Extension!==''&&metadata.Extension!=='0'){
    const target=context.GetObjectBySourceHandle(metadata.Extension);item.ExtensionDictionary=target instanceof api.DxfDictionary?target:null;
    if(item.ExtensionDictionary===null){if(item!==context.Document.Layers||metadata.Extension!==context.layerStateManagerDictionaryHandle||!context.IsAcceptedSourceDictionary(metadata.Extension))throw new FormatException('Unresolved extension dictionary: '+metadata.Extension);}
    else if(item.ExtensionDictionary.Owner!==item)throw new FormatException('Extension dictionary owner mismatch: '+metadata.Extension);
  }
  for(const handle of metadata.Reactors){const target=context.GetObjectBySourceHandle(handle);
    if(target!==null&&!item.PersistentReactors.Contains(target))item.PersistentReactors.Add(target);
    else if(target===null&&handle!=='0'&&!Array.from(context.managedReactorHandles).some(v=>key(v)===key(handle)))throw new FormatException('Unresolved persistent reactor: '+handle);
  }
}
export function ImportDatabaseObjects(context){
  const doc=context.Document,records=context.databaseRecords;
  for(const record of records)context.RecordSourceObject(record.Object,record.SourceIdentity);context.ValidateSourceIdentityDeclarations();
  if(records.Count===0){payload.ResolveSunReferences(context);payload.ResolveOutputSettingsReferences(context);return;}
  const max=0x7fffffffffffffffn;
  for(const record of records){const handle=SourceHandle(record.Object.Handle);if(handle!==null&&handle>=doc.NumHandles&&handle<max)doc.NumHandles=handle+1n;}
  const database=doc.Objects;
  for(const record of records){const item=record.Object;
    const tags=item instanceof api.DxfXRecord?item.Data:item instanceof api.DxfOpaqueObject?item.Tags:
      item instanceof api.DxfStoredTableContent||item instanceof api.DxfStoredTableGeometry||item instanceof api.DxfStoredCellStyleMap||item instanceof api.DxfStoredField||item instanceof api.DxfStoredSunStudy?item.Payload:
      item instanceof api.DxfStoredSectionManager||item instanceof api.DxfTableStyle?item.Tags:[];
    for(const tag of tags)database.ReserveUnresolvedReference(tag);
    for(const data of item.XData.Values)for(const tag of data.XDataRecord)if(tag.Code===api.XDataCode.DatabaseHandle)database.ReserveUnresolvedReference(new api.DxfTag(1005,tag.Value));
  }
  const root=Array.from(records).find(r=>r.Object.Handle===(context.namedDictionary?.Handle??null))??null,managed=new Set();
  if(context.layerStateManagerDictionaryHandle!==null&&context.dictionaries.ContainsKey(context.layerStateManagerDictionaryHandle)){
    const manager=context.dictionaries.get_Item(context.layerStateManagerDictionaryHandle);managed.add(key(manager.Handle));
    for(const child of manager.Entries.Keys){managed.add(key(child));if(context.dictionaries.ContainsKey(child))for(const state of context.dictionaries.get_Item(child).Entries.Keys)managed.add(key(state));}
  }
  if(root!==null){if(!(root.Object instanceof api.DxfDictionary))throw new InvalidCastException();database.ReplaceRoot(root.Object);}
  const collections={ACAD_GROUP:'Groups',ACAD_LAYOUT:'Layouts',ACAD_MLINESTYLE:'MlineStyles',ACAD_IMAGE_DICT:'ImageDefinitions',ACAD_DGNDEFINITIONS:'UnderlayDgnDefinitions',ACAD_DWFDEFINITIONS:'UnderlayDwfDefinitions',ACAD_PDFDEFINITIONS:'UnderlayPdfDefinitions'};
  for(const record of records){if(record===root||managed.has(key(record.Object.Handle)))continue;
    const existing=doc.GetObjectByHandle(record.Object.Handle);
    if(existing!==null){const named={};const collection=record.Object instanceof api.DxfDictionary&&root!==null&&record.Metadata.Owner===root.Object.Handle&&context.namedDictionary.Entries.TryGetValue(record.Object.Handle,named)&&Object.hasOwn(collections,named.value)&&existing===doc[collections[named.value]];
      if(!collection&&!(record.Object instanceof api.DxfXRecord&&existing instanceof api.LayerState))throw new FormatException('Duplicate database identity: '+record.Object.Handle);
      if(collection)context.RecordSourceObject(existing,record.SourceIdentity);managed.add(key(record.Object.Handle));continue;
    }
    database.Register(record.Object,true);
  }
  for(const record of records){if(managed.has(key(record.Object.Handle)))continue;
    if(record!==root&&record.Metadata.Owner!==null&&record.Metadata.Owner!=='0'){
      record.Object.Owner=context.GetObjectBySourceHandle(record.Metadata.Owner);if(record.Object.Owner===null)throw new FormatException('Unresolved database owner: '+record.Metadata.Owner);
    }
  }
  for(const record of records){if(managed.has(key(record.Object.Handle)))continue;const item=record.Object;
    payload.ResolveContainerReferences(context,record);
    if(item instanceof api.DxfDictionary)for(const [name,handle,hard]of record.Entries){
      if(record===root&&api.DxfObjectDatabase.IsReservedName(name))continue;
      const target=context.GetObjectBySourceHandle(handle);if(target===null)throw new FormatException('Unresolved dictionary entry: '+name+' -> '+handle);item.AddLoaded(name,target,hard);
    }
    if(item instanceof api.DxfDictionaryWithDefault&&record.Default!==null&&record.Default!=='0'){
      item.Default=context.GetObjectBySourceHandle(record.Default);if(item.Default===null)throw new FormatException('Unresolved dictionary default: '+record.Default);
    }
    ApplyDatabaseMetadata(context,item,record.Metadata);
  }
  for(const [handle,metadata]of context.entityDatabaseMetadata){const target=context.GetObjectBySourceHandle(handle);if(target!==null)ApplyDatabaseMetadata(context,target,metadata);}
  for(const name of ['ResolveStoredDimAssocReferences','ResolveTableStyleReferences','ResolveStoredTableContentReferences','ResolveStoredTableGeometryReferences','ResolveStoredCellStyleMapReferences','ResolveSunReferences','ResolveStoredFields','ResolveStoredSunStudyReferences','ResolveDataTableReferences','ResolveLayerIndexReferences','ResolveSectionSettingsReferences','ResolveSectionManagerReferences','ResolveDeclaredOwnership','ResolveGeoDataHosts','ResolveOutputSettingsReferences','ResolveLightListReferences'])payload[name](context);
}
